### DISCUSSION TITLE
Welcome to ArangoBnB Discussions!

### DISCUSSION THREAD

**Discussion by cw00dw0rd** [cw00dw0rd/ArangoBnB#1] [General]
## 👋 Welcome! 🥑 
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

## ArangoBnB
The purpose of ArangoBnB is to explore the functionality of ArangoSearch and the various features that it can bring to your applications. This project was created to specifically showcase the ArangoSearch GeoJSON functionality but we would love to have any features added that are relevant to a vacation rental service. Be sure to keep an eye on the projects page for the various tasks and application requirements that we are planning.

**<ins>This project is for educational purposes only.</ins>** 📚 

If you have an idea for a project that you think would be cool to see feel free to start a discussion for it. If you are already working on a project that the community might find interesting please let us know, we would love to share it!

**Comment by lostpebble** [comment]
👋 Hi,

I'm Paul. I have been using ArangoDB in production for almost a year now and loving it so far.

A few years ago, I created a site called [Vibescout](http://www.vibescout.com) which is basically a "things to do" aggregator for South Africans (for the time being). The past year has been... tough 😅 - but we're pushing on cause we're a small team and we've been taking this opportunity to invest the time in new systems, a lot of which make use of ArangoDB on the backend.

So, geo queries are a really big thing for us. I'm very interested to hear about what GeoJSON features are upcoming for ArangoSearch- we found the regular collection geo queries to not meet what we were using before (and continue to use). If we are able to move all our logic into a single database it would be great!

On the other side of things, I use React a lot- so that's where my vote went (but happy to learn other frontend stuff as well). I think I would be able to contribute to the linking of a frontend to a Node.js backend (especially a React frontend)- as that's what I do all day every day, and developed a library that helps a lot with that kind of data management too, [pullstate](https://github.com/lostpebble/pullstate).

Looking forward to seeing where this goes, and contributing where I can!

**Reply by cw00dw0rd** [reply]
Hi Paul,

I am glad to see that you are enjoying ArangoDB and hopefully the geo features will be exactly what you are looking for!
Vibescout looks cool and seems like a great candidate for the geo features of ArangoDB. 

We already have another tutorial that we hope to create in the same way we are doing this one, that one will use whichever framework we don't use for this project. We know that both are very popular and want to have more resources available for as many of the frameworks as we can. So if we don't use React this time it will be next up!

Side note, I enjoy steampunk and coffee and Truth Cafe in Cape Town is on my ever-growing list of travel destinations. 

Thanks again for joining!

**Reply by lostpebble** [reply]
> Side note, I enjoy steampunk and coffee and Truth Cafe in Cape Town is on my ever-growing list of travel destinations.

Ah, awesome- you must definitely try and make a turn that side once we're past this pandemic. Truth is definitely a unique experience, I also really dig the steampunk vibes and they went all out. Amazing coffee too.

> We already have another tutorial that we hope to create in the same way we are doing this one, that one will use whichever framework we don't use for this project. We know that both are very popular and want to have more resources available for as many of the frameworks as we can. So if we don't use React this time it will be next up!

Okay great! Will be interesting to see it from different frontends- though I'm most excited about the power of the new queries at the end of the day ;) I've put so much investment in React, I'm not likely to change any time soon (one man team at the moment, better to stay focused).

> I am glad to see that you are enjoying ArangoDB and hopefully the geo features will be exactly what you are looking for!

Btw, are there any write-ups / blog posts about the new geo features yet? Would be interested to see which direction they're taking.

**Reply by cw00dw0rd** [reply]
> Btw, are there any write-ups / blog posts about the new geo features yet? Would be interested to see which direction they're taking.

Currently, we have the 3.8 ArangoSearch documentation for [Geo Analyzers](https://www.arangodb.com/docs/devel/arangosearch-analyzers.html#geojson) and [Geo Functions](https://www.arangodb.com/docs/devel/aql/functions-arangosearch.html#geo-functions). We will be working on more content as a part of this project as well.


I am always interested to learn about any challenges you might have faced getting started and using ArangoDB. Would you have any interest in doing a community pioneer webinar about Vibescout at some point? Or sharing any challenges (and wins) you faced during your first year in production with ArangoDB?

**Reply by lostpebble** [reply]
Thanks for the links, looks promising. Though on first glance still seems to be using the S2 backend, same as the current geo index. Hopefully the new iteration fixes the performance and other issues we faced with it before. I went into some detail about what problems we faced [here](https://github.com/arangodb/arangodb/issues/11571).

> I am always interested to learn about any challenges you might have faced getting started and using ArangoDB. Would you have any interest in doing a community pioneer webinar about Vibescout at some point? Or sharing any challenges (and wins) you faced during your first year in production with ArangoDB?

Okay cool, I could be open to that. I think I'd have a lot more to share later in the year though. Currently I'm still building a lot of things with ArangoDB which haven't been completely battle-tested or deployed to the production site, at least enough for me to give a complete overview of my thoughts on them. And I'm finally getting my feet wet with graphs etc. as I put in user account things, so I'm sure I'll have a lot more to say in the coming months.

Will try and keep some notes of things that could be of interest while I go :)

**Reply by Simran-B** [reply]
@lostpebble Feel free to share what kind of queries you are interested in particular in case we miss some obvious combinations of features. A high-level description is sufficient, something like "find documents which contain a list of words within a polygon" or "find documents containing a certain phrase within a radius".

**Reply by lostpebble** [reply]
@Simran-B , my biggest request would be the query I was having issues with earlier last year:

For a `POLYGON_A` find all shapes in "geo_index" which INTERSECT the **entire surface area** of `POLYGON_A`, and which include a property `type` equal to `province`.

Previously with ArangoDB, we had to do something like so:

```
FILTER GEO_CONTAINS(polygon_a, doc.geoShape) || GEO_INTERSECTS(polygon_a, doc.geoShape)
FILTER doc.type == "province"
```

This ended up being extremely slow compared to other solutions, because it couldn't make proper use of indexes on the geo functions. But even with attempts at using just one of those geo filters by itself, which employed the index, it was vastly slower than our current solution. Hoping that the ArangoSearch index can improve this kind of query.

My biggest hope is that the new search index includes an index for `GEO_INTERESECTS` that intersects the entire surface area of a polygon, and not just the edges. And that this index is performant, of course.

**Comment by couds** [comment]
Hi all 

I got my eyes in arango from almost 5 years now, I had use it for several small projects but yet to do it on a big one (it's hard to convince investors or temas for non mainstream databases u.u)

I'm also s React developer for frontend so voted for it :( but Im eager to see the new geo index working on the views and the natural language search, to me the full text search had been quite difficult to understand how to correctly implement that part.

Maybe we can do vue and react in parallel to connect to the same backend @lostpebble would you like to?

**Reply by lostpebble** [reply]
Hi @couds , nice to see someone else keen to get involved.

> Maybe we can do vue and react in parallel to connect to the same backend @lostpebble would you like to?

That's actually not a bad idea- I think they wanted to do multiple projects, one for each frontend but with different backend data problems being solved. But having them in parallel here could also work- as you could just define two scripts `npm run react` or `npm run vue` and have them run the same backend but with a different frontend.

@cw00dw0rd and the team would have to decide if that's what they'd like or if encapsulating them separately is preferred.

**Reply by couds** [reply]
Maybe we can do a only api backend and the front can be different projects that way will also be easier for other people to enter and can be used as examples. Wdyt?

**Reply by cw00dw0rd** [reply]
Hi @couds,

Thanks for saying hello and I am glad to see you are enjoying ArangoDB!

Since Vue won the vote that will have to be the primary focus for me, if only due to my available time. However, I think developing a react frontend in parallel is a great idea and if you and @lostpebble would be willing to lead that side of it I am all for it. 

I started with creating a react branch on the repo that you could make PR's against https://github.com/cw00dw0rd/ArangoBnB/tree/react

I have a linked board that I can add you two as collaborators on to make tasks for the community to contribute to as well:
https://github.com/users/cw00dw0rd/projects/1

If that seems like more of a time commitment than anticipated, no worries at all. What do you think?

**Reply by lostpebble** [reply]
@cw00dw0rd , I can't promise I'll always be available time-wise- but will definitely try and put something together. If its just a SPA client, it should be very quick and easy to put together using the "Create React App" tooling. Will try and follow what you decide to do with the Vue frontend. I'm guessing the API backend can be shared between them.

@couds , lets jump on the board and get started :)

**Reply by cw00dw0rd** [reply]
Great, I have added you both to the board. Thanks for helping out with this! I will start adding some actual tasks to the main board today for the backend. Any suggestions/preferences on that?

**Reply by couds** [reply]
@cw00dw0rd Thanks for adding us =)

Sure we can start @lostpebble, if you want we can use [react-bulma-components](https://github.com/couds/react-bulma-components) (Its a small wrapper of bulma.io for React I did) that might come handy.

**Reply by lostpebble** [reply]
@couds - cool, looks like a nice UI library- let's do it. I have my library as well [pullstate](https://github.com/lostpebble/pullstate) which we can use for the data requests :)

> I will start adding some actual tasks to the main board today for the backend. Any suggestions/preferences on that?

@cw00dw0rd I'm a big fan of the [koa framework](https://koajs.com)- I use it for all my projects. But its not much of a big deal which one you go for as I doubt its going to be very complex in any case?

The main thing is that there are API endpoints (which can be created with the `koa-router` package using koa)- which accept GET or POST requests (with a body- can use `koa-body`) and return JSON, with status 200 for success and 401 or whatever for fails- that should allow enough interaction for our frontends.

**Reply by cw00dw0rd** [reply]
@lostpebble KOA seems interesting but now that I think about it, this might be a great opportunity to use [Foxx](https://www.arangodb.com/docs/stable/foxx.html). We will likely only have a few endpoints, no reason to setup a whole other service for that. What do you think?

**Reply by couds** [reply]
@cw00dw0rd I'm not sure about using Foxx for the first example, maybe start with koa or express that have a high amount of users.

This way we do not add too many news things at ones, and we can later create another backend on foxx for comparation.

This way we lower than bar for the entry point. Wdyt?

**Reply by lostpebble** [reply]
@cw00dw0rd I kinda lean towards what @couds is saying- as my usage of ArangoDB is mostly through client libraries and I wouldn't expect to use Foxx as an API endpoint generally- I see its value in executing necessary code and extensions nice and close to the database, but at the end I'd still probably connect to those extension APIs through the Node.js client library, as opposed to the web browser- because of things like session management, logging etc.

But if you think Foxx is poised to take on that role and replace traditional server endpoints altogether, I'm all for giving it a go! Just seems a bit beyond its scope to me at the moment.

**Reply by cw00dw0rd** [reply]
I suspect Foxx would serve our purpose but I also agree that removing barriers is a worthwhile goal, especially when the focus for this project is on ArangoSearch. Let's go ahead with KOA #17 . 

I will make a task for a Foxx service as something to do later, to draw a comparison of the two in a blog post or something.

**Comment by evolvesoft-marwan** [comment]
Hi all, 

I'd been working on a twitter API based project for only a couple of weeks. 

I was initially using MangoDB when I suddenly remembered having come across the concept of graph optimised dbs. I found a nice page which listed all the relevant NoSQL dbs around and I quickly zoomed in on ArangoDB.

So far I managed to insert data into a document collection. Next stop is inserting data into an edge collection e.g. (User1 follows User2).

Btw, I'm using Kafka to deal with streams of data. So had some trouble using the ArangoDB recommended plugin and in the end downloaded the source and compiled the Apache Camel ArangoDB Connect plugin.

I'm sure I'll need to pick your collective brain! So thanks in advance!

**Reply by cw00dw0rd** [reply]
Thanks for saying hi @evolvesoft-marwan and glad to see you are making your way with ArangoDB!
We have a ton of great resources for getting started such as:
 * ArangoDB Learn Portal:
  https://www.arangodb.com/learn/
 * You may find the [AQL](https://www.arangodb.com/docs/stable/aql/tutorial.html) and [Graph Course](https://www.arangodb.com/arangodb-graph-course/) especially useful, they are great starting places.
 * Join our Community Slack channel with 4k+ friendly arangos:
[https://arangodb-community.slack.com/archives/C07NV9NAH](https://arangodb-community.slack.com/archives/C07NV9NAH)

If video courses are more your thing check out our [Getting Started with ArangoDB Udemy course](https://www.udemy.com/getting-started-with-arangodb/).

Always happy to hear about projects you are working on!

**Reply by adhiiisetiawan** [reply]
Hi @cw00dw0rd how to join the Slack channel? i sign in to ArangoDB channel using my email says that "No account for ArangoDB". But usually when i join another workspace just sign in. Is the channel is limited to specific users?

**Reply by cw00dw0rd** [reply]
Hi @adhiiisetiawan 

Sorry that you are having trouble joining the Slack. There shouldn't be any restriction to join. 

The link in the README links directly to the `community-projects` channel so maybe try joining with just the standard ArangoDB Community Slack link: [https://arangodb-community.slack.com/](https://arangodb-community.slack.com/) to see if that works better for you. 

Let me know if you have any more issues trying to join.

**Reply by adhiiisetiawan** [reply]
i still can't join the slack with same issues as before
